torchrl/objectives/value/vtrace.py

```diff
@@ -4,55 +4,334 @@
 
 import torch
+from tensordict.nn import (
+    dispatch,
+    is_functional,
+    set_skip_existing,
+    TensorDictModule,
+    TensorDictModuleBase,
+)
+from tensordict.tensordict import TensorDictBase
+from tensordict.utils import NestedKey
+from torch import nn, Tensor
+from torchrl.objectives.utils import hold_out_net
+from torchrl.objectives.value.advantages import (
+    ValueEstimatorBase,
+    _self_set_skip_existing,
+    _self_set_grad_enabled,
+    _call_value_nets)
+from torchrl.objectives.value.functional import _transpose_time, SHAPE_ERR
 
 
 def _c_val(
     log_pi: torch.Tensor,
     log_mu: torch.Tensor,
     c: Union[float, torch.Tensor] = 1,
 ) -> torch.Tensor:
-    return (log_pi - log_mu).clamp_max(math.log(c)).exp()
-
+    return (log_pi - log_mu).clamp_max(math.log(c)).exp().unsqueeze(-1)  # TODO: is unsqueeze needed?
 
 def _dv_val(
     rewards: torch.Tensor,
     vals: torch.Tensor,
+    next_vals: torch.Tensor,
     gamma: Union[float, torch.Tensor],
-    rho_bar: Union[float, torch.Tensor],
+    rho_thresh: Union[float, torch.Tensor],
     log_pi: torch.Tensor,
     log_mu: torch.Tensor,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
-    rho = _c_val(log_pi, log_mu, rho_bar)
-    next_vals = torch.cat([vals[:, 1:], torch.zeros_like(vals[:, :1])], 1)
-    dv = rho * (rewards + gamma * next_vals - vals)
-    return dv, rho
+    clipped_rho = _c_val(log_pi, log_mu, rho_thresh)
+    deltas = clipped_rho * (rewards + gamma * next_vals - vals)
+    return deltas, clipped_rho
 
 
-def _vtrace(
-    rewards: torch.Tensor,
-    vals: torch.Tensor,
-    log_pi: torch.Tensor,
-    log_mu: torch.Tensor,
-    gamma: Union[torch.Tensor, float],
-    rho_bar: Union[float, torch.Tensor] = 1.0,
-    c_bar: Union[float, torch.Tensor] = 1.0,
+# @_transpose_time # TODO: is this needed?
+def vtrace_correction(
+        gamma: float,
+        log_pi: torch.Tensor,
+        log_mu: torch.Tensor,
+        state_value: torch.Tensor,
+        next_state_value: torch.Tensor,
+        reward: torch.Tensor,
+        done: torch.Tensor,
+        rho_thresh: Union[float, torch.Tensor] = 1.0,
+        c_thresh: Union[float, torch.Tensor] = 1.0,
+        time_dim: int = -2,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
-    T = vals.shape[1]
-    if not isinstance(gamma, torch.Tensor):
-        gamma = torch.full_like(vals, gamma)
-
-    dv, rho = _dv_val(rewards, vals, gamma, rho_bar, log_pi, log_mu)
-    c = _c_val(log_pi, log_mu, c_bar)
-
-    v_out = []
-    v_out.append(vals[:, -1] + dv[:, -1])
-    for t in range(T - 2, -1, -1):
-        _v_out = (
-            vals[:, t] + dv[:, t] + gamma[:, t] * c[:, t] * (v_out[-1] - vals[:, t + 1])
+    """Computes V-Trace off-policy actor critic targets.
+
+    Refer to "IMPALA: Scalable Distributed Deep-RL with Importance Weighted  Actor-Learner Architectures"
+    https://arxiv.org/abs/1802.01561 for more context.
+
+    Args:
+        gamma (scalar): exponential mean discount.
```

Comment:
will this work with gamma as a tensor (per-value gamma)?
If not we should raise an error saying that it is not supported yet and we welcome issues and contributions etc.